force-app/main/default/classes/SOQL.cls

```diff
@@ -287,21 +331,35 @@ public with sharing class SOQL {
      * @return A list of Accounts with matching email domains.
      */
     public static List<Account> findAccountsByLeadEmailDomains(Set<Id> leadIds) {
-        // Query the Leads
-
-        // Create a set of email domains from the Leads
-
-        // Loop through each Lead
-
-            // Get the email from the Lead
-
-            // Check if the email is not null
-
-                // Add the email domain to the set
-
-        
-        // Query the Accounts
-        return null; // Replace null with the variable you used to store the result
+        List<Lead> leads = [SELECT Id, Email FROM Lead WHERE Id IN :leadIds]; 
+
+        Set<String> emailDomains = new Set<String>(); 
+        for (Lead lead : leads) {
+            if (lead.Email != null) {
+                String email = lead.Email.toLowerCase();
```

Comment:
This is broken down nicely and is easy to follow. You could also streamline the code by reducing the domain extraction to a single line
`emailDomains.add('www.' + email.substring(email.indexOf('@') + 1)`